### ISSUE TITLE
Log performance in silent mode

### ISSUE THREAD

**Issue by shivmitra** [j-easy/easy-rules#91]
![easyrules](https://user-images.githubusercontent.com/2118379/26876325-4b61c4a2-4ba3-11e7-9507-2b837b33015a.gif)

I have switched on silent mode.
As you can see, the firerules function takes ~13 seconds, out of which ~6 seconds are spent in various logging functions. 

I was able to shave those 6 seconds off by adding a separate parameter which short circuited all LOGGER. calls. May be the String formatting in logger.info takes time?

**Comment by benas** [comment]
Great! Thank you for sharing these results as requested.

Which version of Easy Rules are you using?

> I was able to shave those 6 seconds off by adding a separate parameter which short circuited all LOGGER. 

What is this parameter? Is it a JVM parameter to disable JUL with a specific properties files and setting level to OFF?

**Comment by shivmitra** [comment]
For now, I just added another engine parameter and then put an if condition before logengineparameter and logrules function. I am using 2.4.0 version.

**Comment by benas** [comment]
It would be a great fail if we are not able to deactivate logging in a non-programmatic way!

Please do me a favor and I'm willing to help if necessary, start your jvm with 
`-Djava.util.logging.config.file="logging.properties"` with `logging.properties` having the following content:

```
handlers=java.util.logging.ConsoleHandler
java.util.logging.ConsoleHandler.level=OFF
``` 

The 6 seconds spent in logging should disappear. If not, please let me know.
I'm not a big fan of programmatically disable logging. And if JUL does not allow it, I'm eager to ditch it!

**Comment by shivmitra** [comment]
I removed withSilentMode and set the logging.properties file as you said, the performance didn't improve though. If you think about it, what you did at withSilentMode where you removed all handlers from your logger, is same as what we are doing with properties file, so it isn't a surprise that this doesn't work.

I had earlier tried replacing your logger with slf4j logger and didn't provide slf4j's simple logger in my pom, it then defaulted to all logging operations to no-op and I had 6 seconds performance improvement. It may be the best way to go with slf4j so that the end user of the library has full control over logging. You might not need silent mode  as engine parameter then.

**Comment by benas** [comment]
Thank you for your time! 6s out of 13s is not acceptable at all, it's about 50%.

I don't exclude removing JUL in favor of SLF4J.

Please give me sometime to experiment with this issue and I'll give you my feedback asap.

Thank you for raising this issue.

**Comment by rongmin** [comment]
i would like to ask how to benchmark this easy rule engine performance

**Comment by benas** [comment]
Hi @rongmin 

This issue is about the performance of the logging framework, not the engine itself.

Easy Rules does not use rete algorithm, it iterates over the rules and executes them. I did a benchmark some time ago (see [here](https://gitter.im/EasyRules/easyrules?at=5511cd4d610f433a6ab0cb4a)), I'm not sure if it is relevant anymore. The best thing to do is to benchmark the engine on your use case.

Kr
Mahmoud

**Comment by benas** [comment]
> It may be the best way to go with slf4j so that the end user of the library has full control over logging. You might not need silent mode as engine parameter then.

@shivmitra We have replaced JUL with SLF4J and deprecated the `silentMode` parameter.
As you said, now the user has full control to choose the logging library.

These changes will be included in the upcoming v3.1 (which you can already test in snapshot version).

Thank you for your valuable feedback!

Kr
Mahmoud
